### PR TITLE
Reopen: Add Black formatting

### PR DESCRIPTION
**Pull request by ethnh**
# Related Issue/Addition to code

- Fixes #
- Code readability issues
- Ensures the repository follows the PEP8 standard
- Remove unnecessary line from CONTRIBUTING.md -- "Please don't make a pull request with a bunch of changes in syntax just for the sake of 'best practices' [...] Unless something makes a difference to performance frankly I don't care." -- There are plenty of PRs that get accepted that do not make the slightest difference to performance, and this line pushes away many from contributing. ( #194 : "So I guess I won't bother. Too bad."  #489 : "I personally don't think this project at this state is salvageable, especially given ThioJoe's attitude." )

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# Proposed Changes
- Add workflows to automatically format and stylize the code using Black
- - This helps to ensure all code is properly formatted with Black, regardless of where it is edited.
- Add .pre-commit-config.yaml for developers with the [pre-commit](https://pre-commit.com/) tool to format their code before they create a pull request.
- Update CONTRIBUTING.md with a note about Black ( Some changes from #489 : @dekoza , and some from #464 's comments : @rachmadaniHaryono ) 
- Adds instructions to setup a development environment

## Checklist:

- [x] My code follows the style guidelines of this project and have read CONTRIBUTING.md
**There are no clear style guidelines. This adds some.**
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any misspellings

# Additional Info

See discussion #194 

Black does **NOT** change any of the underlying code itself, but can help make obvious errors and bugs more visible.
Black does **NOT** fix design problems in the project, but can help highlight areas that need improvement.
Black does **NOT** need to be installed to run the code -- It is only used during development.
If the code is easier to read, it is easier to change.
[Is Black safe to use?](https://black.readthedocs.io/en/stable/faq.html#is-black-safe-to-use)

Pylint Score before Black formatting: -9.18/10
Pylint Score after Black formatting: 0.98/10
The repository would still need big improvements, but a difference of 8.2 is huge.

[Why does this matter?](https://realpython.com/python-code-quality/) 
[Brief tutorial on Pylint](https://pythonspeed.com/articles/pylint/)

This is a reopen of #464 -- I believe the changes in this are valuable enough to be carefully considered, not denied due to a lack of understanding. 
Please **carefully** consider each PR. As you make clear, you may not be an expert, but it does not mean you cannot improve. Contributors are trying to improve the project.
Whether or not Black is the chosen formatter for this project doesn't matter much to me, as long as one is decided upon, and it doesn't cause too much trouble to use.
Similar to testing, readability is important to enforce and will help improve the quality of the code.

Please ignore the failing tests, the codebase in its current state obviously does not adhere to the Black style guidelines, which is where this fail is from.

### CODE REVIEW

**Comment by rachmadaniHaryono** [comment]
some response here https://github.com/ThioJoe/YT-Spammer-Purge/discussions/194#discussioncomment-2118124

should i copy my message here?

**Comment by KendallDoesCoding** [comment]
1c5f35d61eb9fd98b4d8e4fbd88c6b67f7f62b6a The CONTRIBUTING.MD is fine as it is.

**Review by KendallDoesCoding — CHANGES_REQUESTED** [review]
> Tbh I don't really have any intention of adding this, I don't really understand the benefit

Comment by ThioJoe

> Dude chill out, its not your repo. If you don't like a choice you don't have to deal with it lol

Comment by [Masamune3210](https://github.com/Masamune3210)

Duplicate of #464 , ThioJoe, doesnt want to add it, its his project and his wish, so please don't create multiple PR's on the same thing

**Review by KendallDoesCoding — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md
Comment (EthanHindmarsh) [reply CONTRIBUTING.md:14]: What? Why would you not want to emphasize this?

File: CONTRIBUTING.md
Comment (EthanHindmarsh) [reply CONTRIBUTING.md:12]: Why not? Again, emphasis.

File: CONTRIBUTING.md
Comment (KendallDoesCoding) [reply CONTRIBUTING.md:14]: I mean, it's fine as it is, according to me.

File: CONTRIBUTING.md
Comment (EthanHindmarsh) [reply CONTRIBUTING.md:10]: It is not fine as is, my reasoning stated in the PR description:
```Remove unnecessary line from CONTRIBUTING.md -- "Please don't make a pull request with a bunch of changes in syntax just for the sake of 'best practices' [...] Unless something makes a difference to performance frankly I don't care." -- There are plenty of PRs that get accepted that do not make the slightest difference to performance, and this line pushes away many from contributing. ( https://github.com/ThioJoe/YT-Spammer-Purge/discussions/194 : "So I guess I won't bother. Too bad." https://github.com/ThioJoe/YT-Spammer-Purge/pull/489 : "I personally don't think this project at this state is salvageable, especially given ThioJoe's attitude." )```

**Comment by KendallDoesCoding** [comment]
I mean, it's fine as it is, according to me.

**Comment by EthanHindmarsh** [comment]
> I mean, it's fine as it is, according to me.

Sure it's fine, but it could be better.

File: CONTRIBUTING.md
Comment (KendallDoesCoding) [reply CONTRIBUTING.md:10]: Well, ok..

**Comment by KendallDoesCoding** [comment]
> > I mean, it's fine as it is, according to me.
> 
> Sure it's fine, but it could be better.

true

**Comment by EthanHindmarsh** [comment]
> >Tbh I don't really have any intention of adding this, I don't really understand the benefit
>
> Comment by ThioJoe

I believe adding formatting of any sort deserves a second thought, there are plenty of benefits. 
I don't see it as a reason to close the PR as I said in the PR description,
```
This is a reopen of https://github.com/ThioJoe/YT-Spammer-Purge/pull/464 -- I believe the changes in this are valuable enough to be carefully considered, not denied due to a lack of understanding.
Please carefully consider each PR. As you make clear, you may not be an expert, but it does not mean you cannot improve. Contributors are trying to improve the project.
Whether or not Black is the chosen formatter for this project doesn't matter much to me, as long as one is decided upon, and it doesn't cause too much trouble to use.
Similar to testing, readability is important to enforce and will help improve the quality of the code.
```

> > Dude chill out, its not your repo. If you don't like a choice you don't have to deal with it lol
>
> Comment by [Masamune3210](https://github.com/Masamune3210)

This comment was directed to someone else, not towards the PR or myself


> Duplicate of https://github.com/ThioJoe/YT-Spammer-Purge/pull/464 , ThioJoe, doesnt want to add it, its his project and his wish, so please don't create multiple PR's on the same thing

Read the PR description. 
Thio did not say he is against code formatting, only that he did not understand the benefit.

**Comment by KendallDoesCoding** [comment]
> [ @EthanHindmarsh ](/EthanHindmarsh)
> ` [Merge ](/ThioJoe/YT-Spammer-Purge/pull/573/commits/7de772db3228abf166693423a823692d6f9ea443)`

👍

File: CONTRIBUTING.md
Comment (EthanHindmarsh) [reply CONTRIBUTING.md:10]: If a PR is a tiny change with no difference, it will be denied. Ex. Change `is` to `==`
This should be common sense; The line should therefore be redundant, aside from making the project seem hostile to small change.
If one cannot make a big change, and one cannot make a small change, what is one to do?

**Comment by KendallDoesCoding** [comment]
well, i agree but its @thiojoe's project, he can make whatever guidelines
he wants, you've shared your opinion though.

**Comment by ThioJoe** [comment]
I'll consider it, but I'm not going to make a decision either way unless I fully understand it. So no guarantees

**Comment by rachmadaniHaryono** [comment]
is there anything we can do to help you understand it?

**Comment by antjlo** [comment]
You should squash your commits

**Comment by dekoza** [comment]
@antjlo Nope. The squashing is usually done on merge.

**Comment by rachmadaniHaryono** [comment]
just today django just merged pr to autoformat their codebase with black https://github.com/django/django/pull/15387

from that pr there is related deps that explain why they do that https://github.com/django/deps/blob/main/accepted/0008-black.rst

**Comment by dekoza** [comment]
Plus, Black is no longer in beta :smile:

**Comment by showierdata9978** [comment]
Black Is A Official Python Project, I trust it 

GH link: [black](https://github.com/psf/black)

**Comment by VoxelCubes** [comment]
> from that pr there is related deps that explain why they do that https://github.com/django/deps/blob/main/accepted/0008-black.rst

The explanation provided there is very compelling. It saves everyone the headache of thinking about formatting, makes bugs more readily apparent, and makes it easier for new contributes who don't know how the rest of the project was formatted.

**Comment by EthanHindmarsh** [comment]
^^ Ensure no merge conflicts, though I doubt any would exist.

**Comment by rachmadaniHaryono** [comment]
https://github.com/ThioJoe/YT-Spammer-Purge/discussions/194#discussioncomment-2118124

> it does look fine but maybe we can use github action rather than automated commit from an user

from failed github action

```
remote: Permission to [ThioJoe/YT-Spammer-Purge](https://github.com/ThioJoe/YT-Spammer-Purge/).git denied to github-actions[bot].
fatal: unable to access 'ThioJoe/YT-Spammer-Purge': The requested URL returned error: 403
Error: Process completed with exit code 128.
```
something is changed between from github action

**Comment by EthanHindmarsh** [comment]
RE: rachmadani
I don't know too much about GitHub actions (or workflows), but if it does
the same job then LGTM✅

As per the failed GitHub action, it is due to
1. it attempting to automatically push to main when it has not yet received
permission to edit files
2. the repo not being black formatted (causing #1)


I'd like to reiterate,
Any formatting standard would at least set a consistent format for the
code, implementation details don't matter to me😁
The sooner one can be chosen and implemented the better👍

**Comment by rachmadaniHaryono** [comment]
> As per the failed GitHub action, it is due to 

> 1. it attempting to automatically push to main when it has not yet received permission to edit files 
> 2. the repo not being black formatted (causing # 1)

it is number 1

https://github.com/ThioJoe/YT-Spammer-Purge/runs/5225509511?check_suite_focus=true

**Comment by micziz** [comment]
> I'll consider it, but I'm not going to make a decision either way unless I fully understand it. So no guarantees

OK, so basically python gives out standards on how to make your code look readable and under the **Python standard.** One of this is PEP 8. Black is a formmatter witch basically takes your code, format's it for PEP 8 and that's it. Adds no overhead and just and extra command when you need to compile.

**Comment by micziz** [comment]
Also, completely support changing that line in contributing... stooped me from opening a pr once!

**Review by showierdata9978 — APPROVED** [review]
Looks like this is all good black-related stuff.

**Comment by micziz** [comment]
Does this have chances to be added? Even without the removing in contributing.md?

**Comment by EthanHindmarsh** [comment]
> Does this have chances to be added? Even without the removing in contributing.md?

🤷
Unless there are any changes requested or suggested, the PR is ready for review whenever. 
The changes are small enough that there shouldn't be any merge conflicts even after a month

The PR is honestly just waiting for @ThioJoe to make time for a review👍

**Comment by Firecul** [comment]
I've been using this code formatted with Black (manually) for about a month now with no problems. I haven't tried to automate it or test using actions but in principle I completely support this.

**Comment by showierdata9978** [comment]
oh btw it might be time to just request review by theojoe

**Comment by micziz** [comment]
yhea @ThioJoe

**Review by KendallDoesCoding — COMMENTED** [review]
In my opinion, it would be better if the Setting up your Environment section is in README or wiki, I don't think contributing guidelines is the right place for it? I may be wrong.

**Comment by EthanHindmarsh** [comment]
RE: Kendall

Sure, those would probably be better suited to it, but I can only submit
changes to CONTRIBUTING.md 😅
A basic environment setup for new contributors wouldn't hurt though,
wherever it may be

**Comment by KendallDoesCoding** [comment]
> 

You can't move the stuff to README then? or put in the PR description, It would be good if this section gets added to wiki? I mean no harm in CONTRIBUTING.md but it would be better in wiki or readme.

**Comment by showierdata9978** [comment]
Bro this has been open for so long now ,  fully support this, and as i said @ThioJoe look at the repo for it [black](https://github.com/psf/black)

**Comment by KendallDoesCoding** [comment]
> Bro this has been open for so long now , fully support this, and as i said @ThioJoe look at the repo for it [black](https://github.com/psf/black)

@showierdata9978 ThioJoe, is busy, he didn't close it, he still is reviewing it, don't get angry. it's his project bruh

**Comment by micziz** [comment]
I understand that @ThioJoe is busy, but this opened on Feb 5 (even more if we count the last pr)... What is the review process at this?

**Comment by KendallDoesCoding** [comment]
> I understand that @ThioJoe is busy, but this opened on Feb 5 (even more if we count the last pr)... What is the review process at this?

Everyone is complaining about this, ThioJoe is very busy, and for this specific a lot of reviewing has to be done, so just give him some time, the last commit was "14" days ago

**Comment by Firecul** [comment]
I've just started looking through this and experimenting with it on my own fork. I'll prob have a few comments to make tonight

File: .github/workflows/black-push.yml
Comment (Firecul) [inline .github/workflows/black-push.yml:7]: Maybe worth responding to all pushes to catch any possible issues before they hit the main branch?
```suggestion

```

File: .pre-commit-config.yaml
Comment (Firecul) [inline .pre-commit-config.yaml:3]: Exclude confusables from all scans so as to make fetching from upstream easier if @ThioJoe ever wants to update that
```suggestion
exclude: Scripts/confusablesCustom/

repos:
```

File: .pre-commit-config.yaml
Comment (Firecul) [inline .pre-commit-config.yaml:5]:
```suggestion
  rev: v4.3.0
```
Update to the most recent version (fixes an error that can and did crop up in testing)

File: .pre-commit-config.yaml
Comment (Firecul) [inline .pre-commit-config.yaml:10]:
```suggestion
  - id: end-of-file-fixer
  - id: fix-byte-order-marker
  - id: mixed-line-ending
  - id: requirements-txt-fixer
```
If we're using the end-of-file-fixer, there is no harm in fixing more issues in my opinion

File: .pre-commit-config.yaml
Comment (Firecul) [inline .pre-commit-config.yaml:11]:
```suggestion
  - id: trailing-whitespace
    args: [--markdown-linebreak-ext=md]
```
So that `.md` files do not have trailing whitespace trimmed as this could be used for line breaks.

File: .pre-commit-config.yaml
Comment (Firecul) [inline .pre-commit-config.yaml:13]:
```suggestion
  rev: 22.6.0
```
Update to the most recent version

File: .pre-commit-config.yaml
Comment (Firecul) [inline .pre-commit-config.yaml:32]:
```suggestion
  rev: v2.4.0
```
Update to the most recent version

File: .pre-commit-config.yaml
Comment (Firecul) [inline .pre-commit-config.yaml:38]:
```suggestion
    exclude: .github/workflows/codeql-analysis.yml
```
We don't have a `.copier-answers.yml` file but we do have a cron action in the `codeql-analysis.yml` file that gets the `'` removed around asterisks which is not a valid suggestion.

As per [the doc here](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onschedule)
`* is a special character in YAML so you have to quote this string`

<details>  
There are things in the `codeql-analysis.yml` file that should be dealt with but this can be dealt with another pr probably.
</details>

File: .pre-commit-config.yaml
Comment (Firecul) [inline .pre-commit-config.yaml:56]: Also could remove the commented tests but there is no harm in leaving those in.

**Review by Firecul — CHANGES_REQUESTED** [review]
`pre-commit` should also be added into requirements.txt

**Comment by Firecul** [comment]
Sorry for the wall of suggestions @ethnh 😅, I've been diving into this topic to learn and see what's what over the past couple of nights just to make sure I know *something* about this topic.

gitkraken_GitKraken (17)

**Comment by ethnh** [comment]
@Firecul ,

Those changes look great to me👍 I will merge into my branch in a bit when
I get back to my computer 😁
Thanks for taking a look
<https://user-images.githubusercontent.com/34950371/177435486-005156fb-d299-42aa-8f2e-1b70db2f4a9b.png>

File: .github/workflows/black-push.yml
Comment (Firecul) [reply .github/workflows/black-push.yml:7]: I've been using `on: [push, pull_request]` personally to also run on pull requests. That would also be a good option

```diff
-on:
-  push:
-    branches: [ main ]
+ on: [push, pull_request]

Comment (ethnh) [reply .github/workflows/black-push.yml]: Will leave PRs to be formatted on merge, though they should be pre-formatted by a workflow on the branch anyway.

File: .github/workflows/black-push.yml
Comment (ethnh) [reply .github/workflows/black-push.yml:21]: Good catch, `EthanHindmarsh` is my old username too so it wouldn't even point anywhere 😄

File: .pre-commit-config.yaml
Comment (ethnh) [reply .pre-commit-config.yaml:3]: LGTM✅

File: .pre-commit-config.yaml
Comment (ethnh) [reply .pre-commit-config.yaml:12]: Cool looks fine✅

File: .pre-commit-config.yaml
Comment (ethnh) [reply .pre-commit-config.yaml:62]: I don't care much either way, never used poetry or pyupgrade myself

**Comment by Firecul** [comment]
> `pre-commit` should also be added into requirements.txt
https://github.com/ThioJoe/YT-Spammer-Purge/pull/573#pullrequestreview-1027898136


Is the only thing left then https://pre-commit.com/#installation

image

**Comment by KendallDoesCoding** [comment]
> Sorry for the wall of suggestions @ethnh 😅, I've been diving into this topic to learn and see what's what over the past couple of nights just to make sure I know _something_ about this topic.
> 
> gitkraken_GitKraken (17)

No worries, but my email got pretty flooded lol

File: .github/workflows/black-push.yml
Comment (KendallDoesCoding) [reply .github/workflows/black-push.yml]: > Good catch, `EthanHindmarsh` is my old username too so it wouldn't even point anywhere 😄

and your name shouldn't show on it coz this is ThioJoe's project lol

File: CONTRIBUTING.md
Comment (KendallDoesCoding) [inline CONTRIBUTING.md:10]:
```suggestion
- *Please* ensure all code has been formatted by [Black](https://pypi.org/project/black) ([Usage Guide](https://black.readthedocs.io/en/stable/getting_started.html#basic-usage)). This helps to maintain code readability and keep a consistent style across all files. This is optional for you to do.
```

remove extra slash

File: CONTRIBUTING.md
Comment (KendallDoesCoding) [inline CONTRIBUTING.md:22]:
```suggestion
- Download your fork to your computer, either by:
```

File: CONTRIBUTING.md
Comment (KendallDoesCoding) [inline CONTRIBUTING.md:24]:
```suggestion
or by
- - Download as a ZIP and extract [(How?)](https://www.cloudsavvyit.com/p/uploads/2021/11/e8bb2301.png)
```

**Review by KendallDoesCoding — COMMENTED** [review]
just some basic suggestions.

File: CONTRIBUTING.md
Comment (Firecul) [reply CONTRIBUTING.md:10]: That **is** the url though, without the slash autocorrects when entered.

**Comment by ethnh** [comment]
>just some basic suggestions.

@KendallDoesCoding Changes look fine to me ✅

File: CONTRIBUTING.md
Comment (KendallDoesCoding) [reply CONTRIBUTING.md:10]: > That **is** the URL though, without the slash it autocorrects when entered.

wdym?

Comment (Firecul) [reply CONTRIBUTING.md:10]: When you enter ` https://pypi.org/project/black` into the URL bar it is forwarded to `https://pypi.org/project/black/`

image

Therefore I'd suggest that the URL with a `/` is the correct URL as the developers of the site would like.

**Review by Rohith0009 — APPROVED** [review]
LGTM✅

**Pull request closed without merging**
